barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -264,4 +264,141 @@ namespace avm_alu(256);
+    input_ia = (a_lo + 2 ** 128 * a_hi) * cmp_sel;
+
+    pol commit b_lo;
+    pol commit b_hi;
+    // Check input_ib is well formed from its lo and hi limbs
+    input_ib = (b_lo + 2 ** 128 * b_hi) * cmp_sel;
+
+    pol commit p_sub_a_lo; // p_lo - a_lo
+    pol commit p_sub_a_hi; // p_hi - a_hi
+    pol commit p_a_borrow;
+
+    // Check that (p > a) by checking (p_lo > a_lo && p_hi >= ahi) || (p_lo <= a_lo && p_hi > a_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    // (p - 1) lower 128 bits = 53438638232309528389504892708671455232
+    // (p - 1) upper 128 bits = 64323764613183177041862057485226039389 
+    p_sub_a_lo = (53438638232309528389504892708671455232 - a_lo + p_a_borrow * 2 ** 128) * cmp_sel;
+    p_sub_a_hi = (64323764613183177041862057485226039389 - a_hi - p_a_borrow) * cmp_sel;
+
+    pol commit p_sub_b_lo;
+    pol commit p_sub_b_hi;
+    pol commit p_b_borrow;
+
+    // Check that (p > b) by checking (p_lo > b_lo && p_hi >= bhi) || (p_lo <= b_lo && p_hi > b_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    p_sub_b_lo = (53438638232309528389504892708671455232 - b_lo + p_b_borrow * 2 ** 128) * cmp_sel;
+    p_sub_b_hi = (64323764613183177041862057485226039389 - b_hi - p_b_borrow) * cmp_sel;
+
+
+    // Calculate the combined relation: (a - b - 1) * q + (b -a ) * (1-q)
+    // Check that (a > b) by checking (a_lo > b_lo && a_hi >= bhi) || (alo <= b_lo && a_hi > b_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    pol A_SUB_B_LO = a_lo - b_lo - 1 + borrow * 2 ** 128;
+    pol A_SUB_B_HI = a_hi - b_hi - borrow;
+
+    // Check that (a <= b) by checking (b_lo >= a_lo && b_hi >= a_hi) || (b_lo > a_lo && b_hi >= a_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    pol B_SUB_A_LO = b_lo - a_lo + borrow * 2 ** 128;
+    pol B_SUB_A_HI = b_hi - a_hi - borrow;
+
+    pol commit res_lo;
+    pol commit res_hi;
+
+    // If this is a LT operation, we already swapped the inputs so the result of ic is still correct
+    // If this is a LTE operation, we invert the value of ic.
+    pol IS_GT = lt_sel * ic + (1 - ic) * lte_sel;
+
+    res_lo = (A_SUB_B_LO * IS_GT + B_SUB_A_LO * (1 - IS_GT)) * cmp_sel;
+    res_hi = (A_SUB_B_HI * IS_GT + B_SUB_A_HI * (1 - IS_GT)) * cmp_sel;
+
+
+    // ========= RANGE OPERATIONS ===============================
+    // Each call to LT/LTE requires 5x 256-bit range checks. We keep track of how many are left here.
+    pol commit rng_chk_remaining;
+
+    // TODO: combine into 1 equation, left separate for debugging
+    // the number of range checks must decrement by 1 until it is equal to 0;
+    (rng_chk_remaining' - rng_chk_remaining + 1) * rng_chk_remaining = 0;
+    // if this row is a comparison operation, the next range_check_remaining value is set to 4
+    // it is not set to 5 since we do 1 as part of the comparison.
```

Comment:
Well, I think we need to start with 0 for another reason.
The name "rng_check_remaining" is confusing, maybe we should simply name it "ctr" and provide some explanations.

Reply:
We start at 4, so that when we have done the last range check, the ctr or "rng_check_remaining" will be set to 0 and we will know to stop. we actually do 4 additional range checks + the comparison check (which does 1 initial range check)